### DISCUSSION TITLE
Colors as function-like objects

### DISCUSSION THREAD

**Discussion by maxcrees** [typst/typst#2799] [Ideas]
A common pattern I've found when writing my documents in Typst is the following:

```typst
Here is some text with some #text(fill: red)[inline] color.
```

`red` is a predefined color in Typst. We might make our own colors as well:

```typst
#let firebrick = rgb("#b22222")
Here is some text with some #text(fill: firebrick)[inline] color.
```

What I'm wondering is if it's worth it to support using these colors as functions themselves. The semantics to applying them to bare text seems straightforward to me — the above example could be replaced by:

```typst
#let firebrick = rgb("#b22222")
Here is some text with some #firebrick[inline] color.
```

However, I'm not sure what semantics would be desirable for any other elements such color functions would wrap. Would they `#set` all applicable color parameters to that color? Or just some subset? Could this be configurable? What do you think?

**Comment by maxcrees** [comment]
I should add that I know you can implement this yourself of course (untested):

```typst
#let firebrick(body) = text(fill: rgb("#b22222"), body)
Here is some text with some #firebrick[inline] color.
```

But I wonder if there is room for making this a first-class feature.

**Reply by maxcrees** [reply]
The problem with the above approach of course is that you can't make the identifier `firebrick` refer to both the color itself and the function, so you'd need to choose different names.

**Comment by yaksher** [comment]
The main issue that comes to mind with this is that colors are used for a lot of things other than just text color, and so like... what would the behavior of
```typst
#red(block[Test])
```
Would it change just the text `fill` parameter or would it do the `block`'s `fill` parameter too.

**Reply by maxcrees** [reply]
Right, that's what I said above. I'm not sure what behavior would be desired. The most straightforward approach would be to apply it to all color parameters of the child elements.

**Reply by maxcrees** [reply]
Naturally that would make some things unreadable if everything is the same color, but it can be stackable:

```typst
#red[#block[#black[Test]]]
```

**Reply by maxcrees** [reply]
Thinking again about this, it might be more desirable to apply the color to all color-accepting parameters of the top-level child elements. So in your example it would apply to the block's fill & stroke but not to the text itself.
